### ISSUE TITLE
Sqlite Migrations: Table rebuilds

### ISSUE THREAD

**Issue by bricelam** [dotnet/efcore#329]
In SQLite, the following operations will require rebuilding the table - https://docs.microsoft.com/en-us/ef/core/providers/sqlite/limitations.

Other databases also have a smaller set of limitations that require a table rebuild (i.e. changing Identity on a column in SQL Server).

**Comment by bricelam** [comment]
Re-opening. I've regressed this.

**Comment by natemcmaster** [comment]
I've discovered a limitation that makes me wonder if the complication of rebuilds will exceed their value.

The limitation `PRAGMA foreign_keys=on/off;` does not work within a transaction.

Consider the following setup:

Table "Parent" - no FK's
Table "Children" - references "Parent" via a FK

1. Scenario 1 - renaming  "Parent" to "NewTableName". If the `alter table` statement executes with `PRAGMA foreign_keys=ON;`, then SQLite will handle updating the "Children" reference automatically. Otherwise, it will break the FK relationship.
2. Scenario 2 - rebuilding "Parent" ( to imitate drop column or some other operation not natively available). This must be executed with `PRAGMA foreign_keys=OFF;` otherwise the rebuild looks to SQLite like Scenario 1 and "Children" will end up reference a table which is deleted in the rebuild.

Scenario 1 and 2 could be resolved by just executing the proper `PRAGMA foreign_keys` before renaming/rebuilding if not for this limitation,
> This pragma is a no-op within a transaction; foreign key constraint enforcement may only be enabled or disabled when there is no pending BEGIN or SAVEPOINT.

This means we cannot execute an entire migration within one transaction, but need to break SQLite migrations into multiple transactions. Reverting to the beginning of a migration becomes impossible.

**Comment by ErikEJ** [comment]
You are entering a World of Pain :smiley:

**Comment by natemcmaster** [comment]
Moving to backlog to work on high-priority items. 
Consider again after #1141 and #2558.

**Comment by bricelam** [comment]
@natemcmaster Did you try the approach recommended in the SQLite documentation: [Making Other Kinds Of Table Schema Changes](http://sqlite.org/lang_altertable.html#otheralter)? Triggers will be lost. Views may be broken. Indexes would be re-created from the model snapshot. But otherwise it looks like it'll work.

**Comment by bricelam** [comment]
Making this a pri1 issue given the new information.

**Comment by natemcmaster** [comment]
The rebuilds in https://github.com/natemcmaster/EntityFramework/tree/table-rebuild follow a similar pattern. But yes, they can be improved to incorporate the steps mentioned in SQLite. This will require a migration that uses the ADO layer since we can't perform all of those steps in DDL only.

**Comment by bricelam** [comment]
Which part's can't be done? As I see it, the steps for us are:

(Turn off foreign keys during migrations #2592)
(Migrations begins a transaction)

4. Create new table as "new_X"
5. Transfer content from X into new_X
6. Drop the old table X
7. Change the name of new_X to X
8. Reconstruct indexes
9. (Optional) Run `PRAGMA foreign_key_check;`

(Migrations commits transaction)

**Comment by natemcmaster** [comment]
We also need to recreate triggers and views. We would need to detect by querying 'sqlite_master' and storing the results for recreation. This may require altering the definition of the view to reflect the new table structure.

**Comment by bricelam** [comment]
@natemcmaster I think we're ok saying, "Anything you created outside of EF, you're responsible for maintaining." The same is more-or-less true today.

**Comment by bricelam** [comment]
Another scenario to consider when implementing this is changing `IDENTITY` on SQL Server columns. (See #2100)

**Comment by 10Dev** [comment]
I think you are stuck in a "missing the forest for the trees" situation here. Windows 10 UWP should be a "big deal" for Microsoft yet the only EF Data Provider is for SQLite that completely misses out on the EF "Magic Unicorn" of Code First Migrations for any non-trivial data persistance  on UWP. You should have seen this coming down the tracks 3 years ago and now in 2016 a solution for UWP should be the utmost priority. 

The outside world sees just Microsoft, not your internal divisions, so you need to clobber some managers on the head to get ESENT on board, or port SQL Server CE to UWP or start at team to donate fixes to SQLite or point a dedicated team at the awkward "make new table on every migration" solution or whatever gets EF in full glory mode working on UWP.

And it is hard to think "roadmap" on this issue as something should have been in place by the Windows 10 launch and at this point it is just crazy there is no working EF UWP solution.

**Comment by 10Dev** [comment]
I am hoping EF on UWP can become "real" in a rapid timeframe. I was excited to grab the nightly and experience the developer friendlyness of EF only to see it crash and burn when the CodeFirst entities become non-trivial. Then I poked around here and in the docs to discover this insane Catch-22 situation.

So taking my previous comments and translating it into an action list for you to consider from simple to perhaps impossible:

1. Update the documentation again with a more obvious and realistic notice something like "Do not attempt to use EF with the SQLite provider for anything other than classroom demos and the BUILD conference. Real world scenarios will rapidly have you screaming at the EF Team and that will make us sad"

2. Change the label on this Issue from "enhancement" to "bug of the century" and the highest priority and actually assign it to someone. It only looks like an "enhancement" if you see SQLite as the problem instead of seeing it from a customer perspective where the lack of a full EF solution on UWP is the problem.

3. Give bricelam a suitable yeam of great developers and tell him to go crazy on this.

4. investigate the SQL Server team to see if anyone feels competant enough to to the Async thing on CE or whatever embedable chunk of SQL Server goodness they can dream up for UWP.

5. Find out if anyone at Microsoft maintains ESENT and if you can get them to pitch in with whoever is doing the Azure Document DB provider.

6. Get somebody to test various DB's to see if you can find another alternate to SQLite. The BrightstarDB project looks like it might have some UWP potential for example.

7. Throw money at the SQLite organization or start up a Microsoft team to fix their DB.

8. Make up some posters and park yourselves outside Satya's office to point out that "Microsoft's recommended data access technology" needs more resources and separate management from ASP.NET to prioritize resources across all of the Microsoft platforms so that UWP is not just a minor afterthought on ASP.NET

**Comment by 10Dev** [comment]
Rowan has provided excellent feeback on this issue in #4423 and based on that, my wacky solution #7 seems like the most technically feasible approach. With the almost bizarre situation of SQLite now being embedded into Windows 10, the idea of Microsoft donating large amounts of cash and/or other resources to sqlite.org to clean up the mess listed in bricelam's first post on this issue would seem like the ideal alternative to the ugly black hole quagmire that #329 represents. 

Nothing is worse than data loss and this smells like a seed candidate for random impossible to duplicate edge cases of disaster.

But not having a workable UWP EF solution is also leading to the same mess. I don't know about you, but as a developer I poke around looking at user reviews in Windows Store Apps that would be likely to use data such as todo, schedules, etc. It's only anecdotal, but there seems to be an alarming number of users complaining about data loss and my theory on that (which could be tital hogwash of course) is that without EF on UWP (and also missing in action on Windows 8 as well) that most developers have defaulted to object serialaztion of various sorts to plain old files on a disk which as any experienced developer knows is an eventual recipe for data loss. There is a reason we put up with the hassile of ORM to a RDB to get reliable, journalled, transactional data storage. There is no better technology to do this (IMO) than EF. EF missing on UWP = sad face for developers, and possibly horrific disaster stories for users.

If anyone could make #329 work it would have to be the EF team, but that possibly herculean level of effort might just be better spent of fixing SQLite instead.

**Comment by 10Dev** [comment]
Any update on this?

The "enhancement" tag still sits there looking very strange.

The EF Sqlite provider does not work in any real world migration scenario.

The proposed solution is a whole bunch of code that is NOT required for any other database provider and hence seems like some sort of enhancement. But the enhancement is to SQLite really, since it doesn't apply to other databases.

From the point of view of someone using EF, this is a plain and simple bug. 

From a bigger picture, Microsoft only supports SQLite on UWP becomes a meta-bug if EF does not work right.

Is there any possible group inside whoever manages UWP who can be tasked with this effort? Or it can come out of their budget perhaps. Or maybe I'm picking the wrong box to try and think outside of?

Is Core 1.1 possible?

**Comment by rowanmiller** [comment]
@10Dev this isn't currently on our list for 1.1 as there are other higher priority features we want to get thru first.

**Comment by flashfearless** [comment]
Personally, I would think that MS would find it inexcusable to supply EF framework migrations without supporting this functionality.  What is the delay in getting this supported?

**Comment by milasch** [comment]
@flashfearless Well at least it is in 2.0's backlog. :)

**Comment by milasch** [comment]
So I'm trying to add a migration that, by hand, deletes a column on a master table. I created an example of what I'm trying to accomplish by creating 2 model classes:

```
    class Master
    {
        public int Id { get; set; }
        public string ToBeDeleted { get; set;}
    }

    class Detail
    {
        public int Id { get; set; }
        public Master Master { get; set; }
    }

```

I generate the first migration, add data to master and detail. I then delete the ToBeDeleted property then generate the second migration, which creates the infamous migrationBuilder.DropColumn call that isn't supported.

So I replaced that generated script with the manually written one:

```
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.Sql("PRAGMA foreign_keys=\"0\"");
            migrationBuilder.CreateTable(
                name: "NEW_Masters",
                columns: table => new
                {
                    Id = table.Column<int>(nullable: false)
                        .Annotation("Sqlite:Autoincrement", true),
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_Masters", x => x.Id);
                });
            migrationBuilder.Sql("INSERT INTO NEW_Masters SELECT Id FROM Masters;");
            migrationBuilder.DropTable("Masters");
            migrationBuilder.RenameTable("NEW_Masters", newName: "Masters");
            migrationBuilder.Sql("PRAGMA foreign_keys=\"1\"");
        }
```

The problem is that this throws an exception:

> An unhandled exception of type 'Microsoft.Data.Sqlite.SqliteException' occurred in Microsoft.EntityFrameworkCore.Relational.dll
> Additional information: SQLite Error 19: 'FOREIGN KEY constraint failed'.

When I inspect the generated SQL using Migration-Script, it creates SQL that works if I run it manually under SQLiteBrowser, so the exception must be some foreign key constraint check performed by EF.

If I could get some help on a workaround for dropping columns, I would greatly appreciate it.

**Comment by bricelam** [comment]
@milasch Could you [submit a new issue](https://github.com/aspnet/EntityFramework/issues/new) so we can investigate? That all looks correct to me. I wonder if EF is sending a `PRAGMA foreign_keys=1;` when it shouldn't.

**Comment by bricelam** [comment]
Ah yep, as Nate said above:

> `PRAGMA foreign_keys=on/off;` does not work within a transaction.

There is a way to suppress the transaction for a command, but it's only exposed on the `Sql` method. We may need to put together some guidance on how to do this. (cc @rowanmiller)

**Comment by milasch** [comment]
@bricelam No point in submitting a new issue then, right? I'll try to see how I can suppress the transaction.

**Comment by milasch** [comment]
@bricelam The following worked:

        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.CreateTable(
                name: "NEW_Masters",
                columns: table => new
                {
                    Id = table.Column<int>(nullable: false)
                        .Annotation("Sqlite:Autoincrement", true),
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_Masters", x => x.Id);
                });
            migrationBuilder.Sql("INSERT INTO NEW_Masters SELECT Id FROM Masters;");
            migrationBuilder.Sql("PRAGMA foreign_keys=\"0\"", true);
            migrationBuilder.Sql("DROP TABLE Masters", true);
            migrationBuilder.Sql("ALTER TABLE NEW_Masters RENAME TO Masters", true);
            migrationBuilder.Sql("PRAGMA foreign_keys=\"1\"", true);
        }

Thanks for your help!

**Comment by bricelam** [comment]
A cheaper alternative to supporting `DropForeignKey`, `RenameColumn`, and some uses of `AlterColumn` on SQLite would be to use `PRAGMA writable_schema`. (See #7969)

**Comment by kierenj** [comment]
I just got this (`System.InvalidOperationException: 'To change the IDENTITY property of a column, the column needs to be dropped and recreated.'`) when using SQL Server and trying to make an `int Id` into a `ValueGeneratedOnAdd`.  Is this not a bug, rather than an enhancement?  I got no warnings, the migration generated just fine.  But it's invalid?

**Comment by bricelam** [comment]
The migration reflects the desired schema changes. When the migration is applied, it is translated to the appropriate DDL. Applying this schema change requires a table (or in this case, column) rebuild. When this is implemented, the same generated migration will start working because we'll be able to generate the appropriate DML.

The line is always blurry between bugs and enhancements. As a rule of thumb, we bugs tend to be isolated limitations while features are broader and typically involve a significant amount of work. The important thing is that we have an issue tracking the scenario.

**Comment by leak** [comment]
It is really disappointing that such an essential feature as simply renaming a column via migrations is still unsupported in 2.0

**Comment by bricelam** [comment]
@leak Could you be hitting #9344? Renaming columns in Migrations is supported in all providers (e.g. SQL Server, PostgreSQL, etc.) except SQLite which doesn't provide native support for it. Have you considered submitting a feature request on SQLite?

**Comment by leak** [comment]
Well I am hitting most of the things below here: https://github.com/aspnet/EntityFrameworkCore/blob/dev/src/EFCore.Sqlite.Core/Migrations/SqliteMigrationsSqlGenerator.cs#L150

Talking to the SQLite guys seems pointless, they have been evasive about that topic for years now. (see their mailing list archives)

I'm just wondering if SQLite is so unpopular in conjunction with EF, or does it have something to do with exactly those unsupported operations? Am I the only one ever changing table layouts..? I guess not, but why doesn't this issue here doesn't get more traction then? Questions over questions.. 

The SQLite guys outlined what needs to be done for doing proper table rebuilds here: https://sqlite.org/lang_altertable.html

It would really be nice to see a lot of these NotSupportedExceptions disappear and the SQLite EF provider catching up with the other providers.

**Comment by bricelam** [comment]
I agree this is a critical feature for SQLite. I too would like to see it done as soon as possible. Unfortunately, with limited resources, it comes down to a matter of priorities, and we still have some pretty big priorities on our backlog:
* GroupBy translation
* Lazy loading
* Update model from database
* Stored procedure mapping
* Improved view mapping
* Data seeding

...just to name a few.

**Comment by ErikEJ** [comment]
I you need an embedded database on Windows desktop, you could use the SQLCE provider

**Comment by leak** [comment]
@ErikEJ Last time i checked that provider had similar limitations, e.g. not being able to rename columns, or did that change?

**Comment by ErikEJ** [comment]
Yes, rename column is the only Migration limitation in the SQL CE provider, compared to 10 limitations in the SQLite provider - you can work around that with AddColumn and DropColumn

https://docs.microsoft.com/en-us/ef/core/providers/sqlite/limitations

https://github.com/ErikEJ/EntityFramework.SqlServerCompact/wiki/Limitations

**Comment by julielerman** [comment]
Yuck, just hit this today when trying to run a migration that had renamecolumn in it. The renamecolumn method was added by dotnet ef migrations add somechange but can't actually run this migration (rename column not supported).

**Comment by bricelam** [comment]
One less papercut: With #9905, SQLite will now rebuild renamed indexes.

**Comment by leak** [comment]
So I gave this a whirl and from a rough bit of testing it seems to work.
https://github.com/leak/EntityFrameworkCore/commit/f442290cf8ee6d8130e7cf777a2500179af1b8fa

The transaction/PRAGMA issue discussed turned out to be a non issue:
````new SqlOperation { Sql = "PRAGMA foreign_keys=OFF;", SuppressTransaction = true }````
seems to do the trick.

I'm not very familiar with EF code and especially the dependency injection part seems very backwards to me (why not just pass the IoC container around and just let every class grab what it needs..?).
Also it is based on the rel/2.0.0 tag rather than current HEAD.

**Comment by dazinator** [comment]
@bricelam - perhaps worth updating the docs here to reflect rename index support: https://docs.microsoft.com/en-us/ef/core/providers/sqlite/limitations

**Comment by milasch** [comment]
@leak Although I don't understand the use of dependency injection in EF core, passing an IoC container around characterizes a service locator pattern, deemed by many [an anti-pattern](http://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/). But we digress...

**Comment by dazinator** [comment]
> @bricelam - perhaps worth updating the docs here to reflect rename index support: https://docs.microsoft.com/en-us/ef/core/providers/sqlite/limitations

Have submitted a PR for the docs now.

**Comment by leak** [comment]
@milasch Sure, maybe it is just me, but classes like these creep me out big time... https://github.com/aspnet/EntityFrameworkCore/blob/dev/src/EFCore.Relational/Migrations/HistoryRepositoryDependencies.cs

**Comment by WilliamABradley** [comment]
@leak If this fixes all of the issues with SQLite migrations, why don't you submit a PR for it?

**Comment by leak** [comment]
@WilliamABradley I have not received any feedback if this code is even pull-request worthy. Additionally I do not know if this pull request is subject to a CLA signature, which I'm not willing to give.

**Comment by bricelam** [comment]
> I have not received any feedback if this code is even pull-request worthy.

You'll need to sign a CLA before we can review it...

**Comment by bricelam** [comment]
Also, anyone actually interested in submitting a PR shouldn't look at your code since they must agree they...

> ...have sole ownership of the intellectual property rights to my Submission...

**Comment by nbarbettini** [comment]
Adding for consideration: I use the `dotnet new mvc` template (and SQLite) in my [open-source ASP.NET Core book](https://github.com/nbarbettini/little-aspnetcore-book). This limitation means that some messy [workarounds](https://github.com/nbarbettini/little-aspnetcore-book/blob/master/chapters/use-a-database/create-migration.md#workaround-for-sqlite-limitations) are required

**Comment by ajcvickers** [comment]
Make sure to test case in #11832

**Comment by ajcvickers** [comment]
Note: consider case described in #12196

**Comment by bricelam** [comment]
Hey cool, SQLite 3.25.0 supports renaming columns. Filed #13341 to implement it on our side.

**Comment by julielerman** [comment]
Before I dig too deeply I want to see if this is related. This is something I did not experience in 2.0 but am in 2.1.4. I have an existing migration. Then I add in a many to many entity between two entities and add the key mapping for that join entity. When I attempt to add a new migration, it fails. If I delete all of the migrations, I'm able to create a migration with the new mapping. So I believe the attempt of the migration API to create a change constraints is triggering the problem. Is that related to this issue already?

**Comment by julielerman** [comment]
also here's the error : 
 ```
 data git:(TEST_E1) ✗ dotnet ef migrations add init -s ../console
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.EntityFrameworkCore.Update.Internal.SharedTableEntryMap`1.CreateSharedTableEntryMapFactory(IReadOnlyList`1 entityTypes, IStateManager stateManager, String tableName, String schema)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.DiffData(TableMapping source, TableMapping target, DiffContext diffContext)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.Diff(TableMapping source, TableMapping target, DiffContext diffContext)+MoveNext()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.DiffCollection[T](IEnumerable`1 sources, IEnumerable`1 targets, DiffContext diffContext, Func`4 diff, Func`3 add, Func`3 remove, Func`4[] predicates)+MoveNext()
   at System.Linq.Enumerable.ConcatIterator`1.MoveNext()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.Sort(IEnumerable`1 operations, DiffContext diffContext)
   at Microsoft.EntityFrameworkCore.Migrations.Design.MigrationsScaffolder.ScaffoldMigration(String migrationName, String rootNamespace, String subNamespace, String language)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Object reference not set to an instance of an object.
```

**Comment by bricelam** [comment]
@julielerman Looks related to data seeding. Could you [submit a new issue](https://github.com/aspnet/EntityFrameworkCore/issues/new) so we can investigate?

**Comment by bricelam** [comment]
modelBuilder.HasData()

**Comment by julielerman** [comment]
Sorry, my question wasn't clear. Do you mean that even if I'm not using HasData, the seeding APIs are messing this up? Because I'm not using HasData in the context.

**Comment by julielerman** [comment]
Update: After a lot of trial and error (also fails w SQL Server, succeeds on a simple repro,succeeds in 2.0.3 but fails with 2.1.0-2.14) I finally discovered that it succeeds again with the 2.2 preview. So unless you want a proper breadcrumb for this problem which I can't get to the bottom of, I won't bother creating an issue because it seems to be okay going forward. It would be interesting to debug it into internal EF Core code, but probably not worth the time. Let me know if you feel differently about that, @bricelam . Sorry to have bogged down this issue with the side-track.

**Comment by ghost1372** [comment]
is there any news?

**Comment by bricelam** [comment]
@ghost1372 www.microsoft.com/newsapp :trollface:

**Comment by ghost1372** [comment]
> @ghost1372 [www.microsoft.com/newsapp](http://www.microsoft.com/newsapp) :trollface:

:D i mean news about table rebuilds 😋

**Comment by bricelam** [comment]
We're still in the process of planning features for the 3.0 release. I put the `consider-for-next-release` label on it so I can advocate for it like I do every release. The [Entity Framework Core Roadmap](https://docs.microsoft.com/ef/core/what-is-new/roadmap) has some of our early thoughts about the 3.0 release as well as the process we follow when considering features.

**Comment by alexstrickland** [comment]
Xamarin has become an integral part of the development offerings from Microsoft, and therefore it is natural to want to use EF Core on Android and iOS and therefore also use SQLite (what choice?) and yet over 5 years after this issue was opened migrations are still not fully supported. I honestly could not care if rollback is not possible for certain operations or it is inefficient, please just make it seamless and warn about it in the docs.

**Comment by bricelam** [comment]
@alexstrickland In which docs specifically would you like to see it mentioned? It's covered in [SQLite EF Core Database Provider Limitations](https://docs.microsoft.com/ef/core/providers/sqlite/limitations)

**Comment by alexstrickland** [comment]
@bricelam What I mean is please add the functionality to the sqlite migrations to perform the currently unsupported migrations. If they are a bit crummy then perhaps make them opt-in, and document that the migrations are a bit crummy. By crummy I mean that rollback is not possible. You guys have made a glorious migrations system which I don't want to have to hand patch when a change involves an unsupported migration. Particularly as my hand patch will be just as "bad" as what EF will do (no rollback). I get the extra work and no benefit. Am I oversimplifying this?

**Comment by nbelley** [comment]
To add to that, would it be possible at least to generate a throw exception instead of an incorect migration command for the SQLite provider?

**Comment by mauruskuehne** [comment]
We'd love to have this feature as well (even without rollbacks). SQLite with EFCore is a great way to handle local data with Xamarin. But this limitation makes even small changes to the data model hard to implement.

**Comment by nbelley** [comment]
> We'd love to have this feature as well (even without rollbacks). SQLite with EFCore is a great way to handle local data with Xamarin. But this limitation makes even small changes to the data model hard to implement.

I usually just delete all and force a new initial migration... sucks for the user, but it's bullet proof.

**Comment by leak** [comment]
Meanwhile [my little experiment](https://github.com/aspnet/EntityFrameworkCore/issues/329#issuecomment-341780987) is running in production for 1.5 years.

The only issue I can recall is that one time for a more complex change the order of operations that the migration code generator produced was not applicable and needed manually editing (reordering) of the generated migration. Other than that it is running smoothly.

In case that CLA/PR subject comes up again. I do not have the time of bringing that code to MS-grade quality level (cleanup, update, unit tests, etc.), but my code is released under [WTFPL](https://de.wikipedia.org/wiki/WTFPL), meaning I literally do not care the slightest bit how or by whom it is used.

I do have a version updated to 2.2.4 in a private branch and would be willing to share that too in case of interest.

**Comment by bricelam** [comment]
I'm as eager as anyone to get this into EF Core. Unfortunately, it won't be coming in 3.1 as that's a stabilization release and any significant new feature is inherently risky. I promise I'll keep advocating for it.

**Comment by InteXX** [comment]
@nbelley 

> I usually just delete all and force a new initial migration... sucks for the user, but it's bullet proof.

Could you elaborate? What do you do to accomplish this?

**Comment by AdamJachocki** [comment]
Come on, guys. It's been 5 years and still nothing? Really? Is Xamarin and EfCore just a sandbox or real tool for creating real apps? I even started to create my own migration mechanism to work with different RDBMS - also in Sqlite. And it's not so hard. I know that Sqlite has its problems and restrictions, but really. This is just shameful that no one from company like Microsoft isn't able to do this for 5 years.

**Comment by InteXX** [comment]
@nbelley 

> I usually just delete all and force a new initial migration... sucks for the user, but it's bullet proof.

Could you elaborate? What do you do to accomplish this?

**Comment by AndriySvyryd** [comment]
> Come on, guys. It's been 5 years and still nothing? Really? Is Xamarin and EfCore just a sandbox or real tool for creating real apps? I even started to create my own migration mechanism to work with different RDBMS - also in Sqlite. And it's not so hard. I know that Sqlite has its problems and restrictions, but really. This is just shameful that no one from company like Microsoft isn't able to do this for 5 years.

@AdamJachocki There are only 5 people working on EF Core and there's only so much we can do. See https://github.com/aspnet/EntityFrameworkCore/issues/1368#issuecomment-546085835 and https://github.com/aspnet/EntityFrameworkCore/issues/2266#issuecomment-524077785

**Comment by AdamJachocki** [comment]
OK, maybe I overreacted, but I think that this is really strategic functionality, especially if you want to promote Xamarin.

Maybe my ideas would help to push this implementation further.

In my solution (own database migration mechanism), I have:
- fluent part, that creates some kind of universal database schema - tables, indices, foreign keys - as C# objects.

Next I have DatabaseCreator class that receives dialect*, IDbAccess, logger and so on.

Now, dialect. Dialect is a interface expecially built for particular RDBMS. So I have two dialect classes implementing the interface - one for MSSQL, and one for Sqlite.

I won't give you any MSSQL info, because you know it much better, I will give you some Sqlite hints that I use.

First of all, I have list of recreated tables:

`List<string> recreatedTableNames = new List<string>();`

Next I have the most important method:
GetRecreateTableCommand

My mechanism works like that:
DatabaseCreator gets things to do (for example, create table, create index).
Now, every step is taken from dialect like this (for example creating table):

`void CreateTable(ITable table) //ITable has table name collection of IColumn - just schema in C# objects
{
  ExecuteQueryList(dialect.GetCreateTableCommand(table))
}`

ExecuteQueryList just loggs and executes list of queries returned by dialect. Dialect returns mostly really only one query, but may return several queries. So it's better to have a list instead of one query.

Now, let's get back to GetRecreateTableCommand -> this is, as I said, the most important method in Sqlite dialect:

`IList<string> GetRecreateTableCommand(ITable tab, List<string> commonColNames = null)			
{
	string originalName = tab.Name;
	string tempName = "newTempName_" + tab.Name;
	tab.Name = tempName;

	string newTableCreateCommand = string.Join(Environment.NewLine, GetCreateTableCommand(tab));

	string colNames = string.Empty;
	if (commonColNames == null)
		colNames = string.Join(", ", tab.Select(c => c.Name));
	else
		colNames = string.Join(", ", commonColNames);

	tab.Name = originalName;

	return StrToList1($@"PRAGMA foreign_keys=OFF;
					{newTableCreateCommand}
					INSERT INTO {tempName}({colNames}) SELECT {colNames} FROM {originalName}
					DROP TABLE {originalName};
					ALTER TABLE {tempName} RENAME TO {originalName};

					PRAGMA foreign_key_check;
					PRAGMA foreign_keys=ON;");				
}`

So this method just does what Sqlite team tells to - copies the table to temp table, drops the original table and recreates it. Todo -> copy data back to recreated table.

So for example if you have to change column type name on a table, it's just:
`
public IList<string> GetAlterColumnChangeTypeCommand(ITable tab, IColumn col)
{
	if (recreatedTableNames.Contains(tab.Name))
		return null;

	recreatedTableNames.Add(tab.Name);
	return GetRecreateTableCommand(tab);
}

`
So the thing is:
- user by fluent api creates database schema
- database creator does needed changes
- as far as sqlite is concerned, most changes are just - recreate the table. So, when we have to drop foreign key, whole table is recreated using the new schema, and then this table name is being added to helper list (recreatedTableNames). 

Now, some Sqlite queries that are helpful:

`
public string GetColumnExistsCommand(string tableName, string columnName)
{
	return $@"select count(*) as result
			from pragma_table_info('{tableName}')
			WHERE name = '{columnName}'";
}

//this one is really helpful. It gets most of table information as you can get.

public string GetColumnsInfoCommand(string tableName)
{
	return $@"select
			ti.name as Name,
		    ti.type as TypeName,
		    ti.dflt_value as DefaultValue,
		    (case when ti.dflt_value IS NULL THEN 0 ELSE 1 END) as HasDefaultValue,
		    ti.pk as IsPrimaryKey,
		    (case when ti.[notnull] = 1 THEN 0 ELSE 1 END) as Nullable,
		    (case when indexInfo.indexName IS NULL THEN 0 ELSE 1 END) as IsIndex,
		    (case when indexInfo.IndexDescending = 0 then 1 else 0 end) as IndexAscending,
		    (case when fkInfo.[ownerColName] IS NULL THEN 0 ELSE 1 END) as IsForeignKey,
		    fkInfo.foreignTableName as ForeignTableName,
		    fkInfo.foreignColName as ForeignColumnName,
		    (case 
			    when fkInfo.on_update = 'CASCADE' THEN 'Cascade' 
			    when fkInfo.on_update = 'SET DEFAULT' THEN 'SetDefault'
			    when fkInfo.on_update = 'SET NULL' THEN 'SetNull'        
			ELSE 'NoAction' END) as ForeignKeyOnUpdateAction,
			(case 
			    when fkInfo.on_delete = 'CASCADE' THEN 'Cascade' 
			    when fkInfo.on_delete = 'SET DEFAULT' THEN 'SetDefault'
			    when fkInfo.on_delete = 'SET NULL' THEN 'SetNull'        
			ELSE 'NoAction' END) as ForeignKeyOnDeleteAction
		from pragma_table_info('{tableName}') as ti
		left join
		(
			select m.name as indexName, ii.name as colname, ii.[desc] as IndexDescending
			FROM sqlite_master as m
			left join pragma_index_list('{tableName}') pil on pil.name = m.name
			left join pragma_index_xinfo(m.name) as ii on ii.name IS NOT NULL
		
			where m.type = 'index' and m.tbl_name = '{tableName}' and pil.origin<> 'u' and pil.origin<> 'pk'
		) as indexInfo on indexInfo.colname = ti.name
		left join
		(
			select[table] as foreignTableName, [from] as ownerColName, [to] as foreignColName, on_update, on_delete
			from pragma_foreign_key_list('{tableName}')
		) as fkInfo on fkInfo.ownerColName = ti.name";
}


//And my columnInfo class is just:

public class BasicColInfo
{
	public string Name { get; set; }
	public string TypeName { get; set; }
	public object DefaultValue { get; set; }
	public bool HasDefaultValue { get; set; }
	public bool IsPrimaryKey { get; set; }
	public bool Nullable { get; set; }
	public bool IsIndex { get; set; }
	public bool IndexAscending { get; set; }
	public bool IsForeignKey { get; set; }
	public string ForeignTableName { get; set; }
	public string ForeignColumnName { get; set; }
	public ForeignKeyRule ForeignKeyOnUpdateAction { get; set; }
	public ForeignKeyRule ForeignKeyOnDeleteAction { get; set; }
}

public string GetTableExistsCommand(string tableName)
{
	return $@"SELECT count(*) FROM sqlite_master WHERE type = 'table' AND name = '{tableName}'";
}
`

I have also ColumnComparer class that is used to detect schema alterations and update the database schema. I think you have much more better mechanism for this, but this is mine.

I check if the table exists - if yes, there may be alteration required.
Now I check columns in the table, using query from GetColumnsInfoCommand and create list of ColumnInfo class objects.

Next, I just compare it like that:

`
enum ColumnCompareResult
{
	ChangeNullable,
	ChangeDefaultValue,
	ChangePrimaryKey,
	ChangeType,
	ChangeIndex,
	ChangeForeignKey
}

class ColumnComparer
{

	IDbDialect dialect;

	public ColumnComparer(IDbDialect dialect)
	{
		this.dialect = dialect;
	}

	bool NullableChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		return requiredSchemaColumn.CanBeNull != existingColumn.Nullable;
	}

	bool DefaultValueChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		return requiredSchemaColumn.DefaultValue != existingColumn.DefaultValue;
	}

	bool PrimaryKeyChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		return requiredSchemaColumn.IsIdentity != existingColumn.IsPrimaryKey;
	}

	bool FieldTypeChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		string sqlFieldType = dialect.GetFieldTypeFromDbType(requiredSchemaColumn.ColType, requiredSchemaColumn.Length);
		if (string.Compare(sqlFieldType, existingColumn.TypeName, true) != 0)
			return true;
		else
			return false;
	}

	bool IsIndexChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		return (!(requiredSchemaColumn.IndexInfo != null && existingColumn.IsIndex));
	}

	bool IndexSortChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		if (requiredSchemaColumn.IndexInfo == null)
			return false;


		IndexSortOrder existingSortOrder = IndexSortOrder.Descending;
		if (existingColumn.IndexAscending)
			existingSortOrder = IndexSortOrder.Ascending;

		return requiredSchemaColumn.IndexInfo.SortOrder != existingSortOrder;
	}

	bool IndexChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		return IsIndexChanged(requiredSchemaColumn, existingColumn) || (IndexSortChanged(requiredSchemaColumn, existingColumn));
	}


	bool IsFkChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		return (!(requiredSchemaColumn.ForeignKeyInfo != null && existingColumn.IsForeignKey));
	}

	bool FkKeyRuleChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		if (requiredSchemaColumn.ForeignKeyInfo == null)
			return false;

		return requiredSchemaColumn.ForeignKeyInfo.UpdateRule != existingColumn.ForeignKeyOnUpdateAction ||
			requiredSchemaColumn.ForeignKeyInfo.DeleteRule != existingColumn.ForeignKeyOnDeleteAction;
	}

	bool FkTableChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		if (requiredSchemaColumn.ForeignKeyInfo == null)
			return false;

		return string.Compare(requiredSchemaColumn.ForeignKeyInfo.ForeignTableName, existingColumn.ForeignTableName, true) != 0;
	}

	bool FkColumnChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		if (requiredSchemaColumn.ForeignKeyInfo == null)
			return false;

		return string.Compare(requiredSchemaColumn.ForeignKeyInfo.ForeignTableName, existingColumn.ForeignTableName, true) != 0;
	}

	bool ForeignKeyChanged(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		return IsFkChanged(requiredSchemaColumn, existingColumn) ||
			FkKeyRuleChanged(requiredSchemaColumn, existingColumn) ||
			FkTableChanged(requiredSchemaColumn, existingColumn) ||
			FkColumnChanged(requiredSchemaColumn, existingColumn);
	}

	public List<ColumnCompareResult> CompareColumns(IColumn requiredSchemaColumn, BasicColInfo existingColumn)
	{
		List<ColumnCompareResult> result = new List<ColumnCompareResult>();

		if (NullableChanged(requiredSchemaColumn, existingColumn))
			result.Add(ColumnCompareResult.ChangeNullable);

		if (DefaultValueChanged(requiredSchemaColumn, existingColumn))
			result.Add(ColumnCompareResult.ChangeDefaultValue);

		if (PrimaryKeyChanged(requiredSchemaColumn, existingColumn))
			result.Add(ColumnCompareResult.ChangePrimaryKey);

		if (FieldTypeChanged(requiredSchemaColumn, existingColumn))
			result.Add(ColumnCompareResult.ChangeType);

		if (IndexChanged(requiredSchemaColumn, existingColumn))
			result.Add(ColumnCompareResult.ChangeIndex);

		if (ForeignKeyChanged(requiredSchemaColumn, existingColumn))
			result.Add(ColumnCompareResult.ChangeForeignKey);

		return result;
	}
}
`

So, I hope I could help and that Sqlite will be better serviced in EfCore.

**Comment by InteXX** [comment]
@AdamJachocki 

Thank you for the comprehensive code samples. Could you edit your post and set the markdown formatting for snippets, i.e. correct the indentation so that all the code is displayed as code? Thanks.

**Comment by AdamJachocki** [comment]
> 
> 
> @AdamJachocki
> 
> Thank you for the comprehensive code samples. Could you edit your post and set the markdown formatting for snippets, i.e. correct the indentation so that all the code is displayed as code? Thanks.

Done

**Comment by InteXX** [comment]
Very good, thank you.

This appears surprisingly simple. Hopefully, the team will be able to consider using some of these ideas.

**Comment by leak** [comment]
Glad to see that I'm not alone in the SQLite+EF woods, I was always wondering by the lack of activity in this issue..

So I updated my code from above to 2.2.6 [1], see here: https://github.com/leak/EntityFrameworkCore/commit/71d3a35cb100bcdbd84eee362d49385134b28126

I also added the nuget packages that I am using in my project here: https://github.com/leak/EntityFrameworkCore/releases/tag/2261



Few notes:
- The packages are drop-in replacements for the official 2.2.6 packages, just throw them in a folder, create a nuget.config and point them to that folder, that's it
- My tablerebuild226 branch is based on the release commit for the 2.2.6 nuget packages
- I changed the .UseSqlite() method to .UseSqliteX() to really make sure my code is used and I wouldn't get screwed by package caches
- The 2.2.6.1 nuget packages linked above are pure release builds from my branch, with just the version number edited, again to avoid package caching issues

And the usual disclaimer: This code works for me in production, but there still might be dragons. Again my code is released under ([WTFPL](https://en.wikipedia.org/wiki/WTFPL)), use it, leave it, make a PR in this repo - I simply do not care.



[1] Since MS decided to drop full framework support for anything 3.x (yay..) I did not really attempt to run this on anything beyond 2.x

**Comment by TomasHubelbauer** [comment]
I'm also very much interested in SQLite + EF Core! Do not have anything substantial to add, just excited to see a hint of activity on this topic.

**Comment by ErikEJ** [comment]
Sqlite engine supports column rename now, no need for rebuild

**Comment by InteXX** [comment]
@ErikEJ 

This is very good news. Thank you for keeping up with their progress on this.

**Comment by bricelam** [comment]
Note, we added support for column renames in EF Core 2.2

**Comment by 1Mgtrn3** [comment]
@leak 
Thank you! It worked, although it may also require to install Microsoft.CodeAnalysis.CSharp.Workspaces. 

I had an issue "Could not get the reflection type for dbcontext..." when I tried to scaffold a controller using a DbContext added with UseSqliteX(). Installing Workspaces package resolved that issue.
I can say that the issue was caused by installing that custom package because I had scaffolded a couple of controllers successfully in that project before changing Sqlite.Core to a custom version.

**Comment by leak** [comment]
@1Mgtrn3 Thanks for the feedback.

To be honest I never used scaffolding, so since I changed a public api there there is a chance it messed with that.

Just out of curiosity, did you try renaming that method back and see if the issue goes away?

**Comment by jedidiah81** [comment]
if a table cant be altered - ie migration fails - is there anyway of forcing it to recreate the tables (even if that loses data) , rather than having to recreate the whole database and lose everything?

**Comment by gojanpaolo** [comment]
@jedidiah81 you could add a `try-catch` in your `Up` method to do that

**Comment by Epstone** [comment]
Just as an information how to generate sql code for the migration with some tooling support:

Download sqlite studio, do your changes manually in the schema editor until your migration is mirrored there. Before commit you can export the created sql code for migration and merely copy and paste that as replacement for the migration steps. 

'Down' migration may be done the other way around.

Still error prone and code needs to be reviewed and tested, but less pain for writing table rebuilds again and again. Maybe it helps some one right now

**Comment by bricelam** [comment]
Nothing really new here, just jotting down some notes sine I started thinking about this again today:

``` csharp
// UNDONE: Not supported by SQLite
//migrationBuilder.AlterColumn<string>(
//    name: "Title",
//    table: "Posts",
//    nullable: false,
//    oldClrType: typeof(string),
//    oldType: "TEXT",
//    oldNullable: true);

// Create a new table with the desired schema
// TODO: Randomize the name to avoid conflicts
migrationBuilder.CreateTable(
    name: "new_Posts",
    columns: table => new
    {
        Id = table.Column<int>(nullable: false)
            .Annotation("Sqlite:Autoincrement", true),
        Title = table.Column<string>(nullable: false)
    },
    constraints: table =>
    {
        table.PrimaryKey("PK_Posts", x => x.Id);
    });

// Copy data from the old table. Use NULLIF to specify a default value for newly
// required columns
migrationBuilder.Sql(@"
    INSERT INTO new_Posts (Id, Title)
    SELECT Id, IFNULL(Title, '')
    FROM Posts;
");

// Suspend foreign key enforcement during the swap
// TODO: Can't do this on SQL Server. Would need to rebuild referencing foreign
// keys there. But do we even need table rebuilds on SQL Server? Changing
// IDENTITY only requires a column rebuild
// NB: This commits the current transaction. We can't rollback the migration if
// anything after this fails. Maybe we can turn this off before the migration
// and somehow use PRAGMA foreign_key_check. Otherwise, we can mitigate it by
// doing rebuilds as late as possible
migrationBuilder.Sql("PRAGMA foreign_keys = 0;", suppressTransaction: true);

// Swap in the new table
migrationBuilder.DropTable(
    name: "Posts");
migrationBuilder.RenameTable(
    name: "new_Posts",
    newName: "Posts");

// TODO: We shouldn't do this if foreign key enforcement was off to begin with.
// There's no way to handle this in SQL, so we probably need a way to configure
// this
migrationBuilder.Sql("PRAGMA foreign_keys = 1;", suppressTransaction: true);

// Rebuild any indexes
migrationBuilder.CreateIndex(
    name: "IX_Posts_Title",
    table: "Posts",
    column: "Title");
```

**Comment by bricelam** [comment]
I looked into using `PRAGMA defer_foreign_keys` to avoid `suppressTransaction: true`, but dropping the table sill cascade-deletes dependents.

We discussed the consequences of `suppressTransaction: true` in a design meeting today and decided that it is OK to do this by default on SQLite. We did this for certain scenarios in EF6 on SQL Server and never heard any negative feedback. An option to disable it (and thus table rebuilds) would, however, be nice to have.

We also discussed blindly turning foreign keys back on (when we don't know if they were enabled in the first place) and decided that it's OK. If you turn them off for performance reasons, your data should still be consistent otherwise other things in EF will break. The default foreign key behavior will just be reset when the connection is closed and re-opened after migrating anyway.

Other points that came up in the discussion:
* We should warn when we do a rebuild that any changes you've made outside of your EF model may be lost. (But this shouldn't really be an issue if you're using migraitons to manage your database schema)
* There's no value in re-using parts of the query or update pipeline to generate the INSERT INTO..SELECT statements. Just use the usual mechanism for generating SQL in migrations
* No need to support table rebuilds on SQL Server as part of this. File a new issue for doing column rebuilds when IDENTITY changes

**Comment by danielmeza** [comment]
From the SQLiteStudio there is a short sintax to clone the original table
```SQL
CREATE TABLE sqlitestudio_temp_table AS SELECT *
                                          FROM Table;
```
this way we can reduce the complexity.

**Comment by ir2000** [comment]
@danielmeza you would lose primary key and foreign keys 
> A table created using CREATE TABLE AS has no PRIMARY KEY and no constraints of any kind. The default value of each column is NULL. The default collation sequence for each column of the new table is BINARY.
https://www.sqlite.org/lang_createtable.html#createtabas

**Comment by InteXX** [comment]
@ir2000 @danielmeza 

Yes, in that single statement data is lost. But it's a valuable technique for cloning the schema.

One could then clear out the data and repopulate, bringing all keys along this time.

**Comment by danielmeza** [comment]
@ir2000 yes but we not need that info in the temp table.

**Comment by danielmeza** [comment]
@bricelam we should cover in this version escenarios with an existing db? Example: code first from db escenarios.

**Comment by Damus765** [comment]
@danielmeza I would say we need to create every constraint in the "temp table". Because what you call "temp table" is not really temporary. It is in fact the new table that will be kept once renamed. So I'm not sure CREATE TABLE AS is the best way to do create the new table.

**Comment by danielmeza** [comment]
@Damus765 I was missing the rename process, I thought that the data will be copy to the new table so the process goes like this. Create the temp table -> delete the old table -> create the new table with the new schema -> copy data back from the temp table -> delete the temp table.

**Comment by Damus765** [comment]
According to the [SQLite documentation](https://sqlite.org/lang_altertable.html#otheralter), the recommended steps are as follows:

1. Create new table
2. Copy data from old table into new table
3. Drop old table
4. Rename new into old

Only one data copy is required and you don't need a temp table.

**Comment by bricelam** [comment]
@danielmeza Our strategy for using migrations with existing databases is tracked by #2167

**Comment by bricelam** [comment]
More related issues for SQL Server: https://github.com/dotnet/efcore/issues/12586, https://github.com/dotnet/efcore/issues/16758 & https://github.com/dotnet/efcore/issues/20943

**Comment by bricelam** [comment]
![it's been 84 years...](https://media.giphy.com/media/13mPeKqo7YBZhS/giphy.gif)

But seriously, my six-year-old wasn't even born when I filed this issue. Crazy.

May all your rebuilding dreams come true! Check out our [daily builds](https://github.com/dotnet/aspnetcore/blob/master/docs/DailyBuilds.md) if you're eager to give it a try. [Let me know](https://github.com/dotnet/efcore/issues/new?labels=customer-reported&template=bug_report_efcore.md) if you find any bugs. (I still have a bit more testing to do myself.)
